crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1172,14 +1172,36 @@ impl<'db> Type<'db> {
                 // TODO: see above
                 Truthiness::Ambiguous
             }
-            Type::Instance(InstanceType { class }) => {
-                // TODO: lookup `__bool__` and `__len__` methods on the instance's class
-                // More info in https://docs.python.org/3/library/stdtypes.html#truth-value-testing
-                // For now, we only special-case some builtin classes
+            instance_ty @ Type::Instance(InstanceType { class }) => {
                 if class.is_known(db, KnownClass::NoneType) {
                     Truthiness::AlwaysFalse
                 } else {
-                    Truthiness::Ambiguous
+                    // We only check the return value of __bool__ method for truth testing.
+                    // Since the __bool__ method can take priority over __len__ we cannot rely on
+                    // the value of __len__ alone. One example is when a class has a __len__ that
+                    // is 0 but a subclass overrides __bool__ in this case the subclass truth
+                    // testing should be true.
+                    let bool_method = instance_ty.to_meta_type(db).member(db, "__bool__");
+                    // Check if the class has `__bool__ = bool` in the definition and avoid
+                    // recursion
+                    if let Symbol::Type(Type::ClassLiteral(ClassLiteralType { class }), _) =
+                        bool_method
+                    {
+                        if class.name(db) == "bool" {
+                            return Truthiness::Ambiguous;
+                        }
+                    }
```

Comment:
I wonder if we could do this inside `call_dunder` instead, so as to avoid the duplicated `to_meta_type...member` calls, which `call_dunder` will end up doing again.

Reply:
After looking at it, I decided that's no better; this special case is a bit ugly no matter where it goes, and this is probably the best place for it.

I did rework the condition slightly to ensure that we really do have the `bool` built-in and not some random other class named `bool`.